### DISCUSSION TITLE
Practical capabalities of coix?

### DISCUSSION THREAD

**Discussion by Chengwei94** [jax-ml/coix#41] [General]
Hi there, coix seems like a nice package that I am interested in exploring for biological purposes. I must admit I have only skimmed through the paper.  However, from the first look, I cant wrap my head around why I should use coix over functionalities that are already in numpyro. What benefits do these inference combinators bring over just for example, using SVI or MCMC from numpyro?

**Comment by fehiepsi** [comment]
Hi @Chengwei94, coix is more suitable for variational inference than mcmc (though you can use mcmc via the `NUTS(potential_fn=...)` api). Inference combinators allow you to combine multiple programs together, which typically happens in variational, sequential important, or sequential monte carlo settings. In the typical variational setting, you have two programs `model` and `guide`, you can use the api to apply inference objectives like reweighted wake-sleep (which is suitable for models with non-reparameterized variables, like discrete latent variables).

**Reply by zmheiko** [reply]
To add a little to what @fehiepsi said, while you can run standard VI and MCMC just fine using the coix api it does not really provide you with any advantage over using numpyro or any other implementation. The strength of the coix api (and underlying abstraction) is that it provides a neat way of specifying importance-sampling based algorithms (e.g. SMC) and to train them variationally by composing multiple models and guides with the provided program transformations.

**Reply by Chengwei94** [reply]
Thanks for the reply! I think I kind of understand what problems that coix are more suitable for dealing. I think the docs can benefit from an example where SVI or NUTs fails but coix performs well. That would guide practitioners like me to adopt coix in their workflow.
